impeller/renderer/backend/vulkan/render_pass_builder_vk.cc

```diff
@@ -119,57 +117,59 @@ vk::UniqueRenderPass RenderPassBuilderVK::Build(
     color_attachments_count++;
   }
 
-  std::vector<vk::AttachmentDescription> attachments;
-
-  std::vector<vk::AttachmentReference> color_refs(color_attachments_count,
-                                                  kUnusedAttachmentReference);
-  std::vector<vk::AttachmentReference> resolve_refs(color_attachments_count,
-                                                    kUnusedAttachmentReference);
+  std::array<vk::AttachmentDescription, kMaxAttachments> attachments;
+  std::array<vk::AttachmentReference, kMaxColorAttachments> color_refs;
+  std::array<vk::AttachmentReference, kMaxColorAttachments> resolve_refs;
   vk::AttachmentReference depth_stencil_ref = kUnusedAttachmentReference;
+  size_t attachments_index = 0;
+  size_t color_index = 0;
+  size_t resolve_index = 0;
 
   if (color0_.has_value()) {
     vk::AttachmentReference color_ref;
-    color_ref.attachment = attachments.size();
+    color_ref.attachment = attachments_index;
     color_ref.layout = vk::ImageLayout::eGeneral;
-    color_refs[0] = color_ref;
-    attachments.push_back(color0_.value());
+    color_refs[color_index++] = color_ref;
```

Comment:
I suggest using `.at(color_index++)` instead of `[color_index++]`. The former performs range checking in debug for throwing the exceptions (which we disable). The latter is free to write into garbage memory. Some safety net in debug should be good to have IMO.

Reply:
I can add a break if we hit more than 16 attachments. otherwise there shouldn't (TM) need to be bounds checks.

Reply:
added at

Reply:
There are other places where the bounds checks would be harder to add so I made them use .at() too.